platforms/platforms_test.go

```diff
@@ -0,0 +1,233 @@
+		},
+		{
+			input: "armel",
+			expected: specs.Platform{
+				OS:           defaultOS,
+				Architecture: "arm",
+				Variant:      "v6",
+			},
+			formatted: "linux/arm/v6",
+		},
+		{
+			input: "armhf",
+			expected: specs.Platform{
+				OS:           defaultOS,
+				Architecture: "arm",
+			},
+			formatted: "linux/arm",
+		},
+		{
+			input: "Aarch64",
+			expected: specs.Platform{
+				OS:           defaultOS,
+				Architecture: "arm64",
+			},
+			formatted: joinNotEmpty(defaultOS, "arm64"),
+		},
+		{
+			input: "x86_64",
+			expected: specs.Platform{
+				OS:           defaultOS,
+				Architecture: "amd64",
+			},
+			formatted: joinNotEmpty(defaultOS, "amd64"),
+		},
+		{
+			input: "Linux/x86_64",
+			expected: specs.Platform{
+				OS:           "linux",
+				Architecture: "amd64",
+			},
+			formatted: "linux/amd64",
+		},
+		{
+			input: "i386",
+			expected: specs.Platform{
+				OS:           defaultOS,
+				Architecture: "386",
+			},
+			formatted: joinNotEmpty(defaultOS, "386"),
+		},
+		{
+			input: "linux",
+			expected: specs.Platform{
+				OS:           "linux",
+				Architecture: defaultArch,
+			},
+			formatted: joinNotEmpty("linux", defaultArch),
+		},
+		{
+			input: "s390x",
+			expected: specs.Platform{
+				OS:           defaultOS,
+				Architecture: "s390x",
+			},
+			formatted: joinNotEmpty(defaultOS, "s390x"),
+		},
+		{
+			input: "linux/s390x",
+			expected: specs.Platform{
+				OS:           "linux",
+				Architecture: "s390x",
+			},
+			formatted: "linux/s390x",
+		},
+		{
+			input: "macOS",
+			expected: specs.Platform{
+				OS:           "darwin",
+				Architecture: defaultArch,
+			},
+			formatted: joinNotEmpty("darwin", defaultArch),
+		},
+	} {
+		t.Run(testcase.input, func(t *testing.T) {
+			if testcase.skip {
+				t.Skip("this case is not yet supported")
+			}
+			p, err := Parse(testcase.input)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if !reflect.DeepEqual(p, testcase.expected) {
+				t.Fatalf("platform did not match expected: %#v != %#v", p, testcase.expected)
+			}
+
+			formatted := Format(p)
+			if formatted != testcase.formatted {
+				t.Fatalf("unexpected format: %q != %q", formatted, testcase.formatted)
+			}
+
+			// re-parse the formatted output and ensure we are stable
+			reparsed, err := Parse(formatted)
+			if err != nil {
+				t.Fatalf("error parsing formatted output: %v", err)
+			}
+
+			if Format(reparsed) != formatted {
+				t.Fatalf("normalized output did not survive the round trip: %v != %v", Format(reparsed), formatted)
+			}
+		})
+	}
+}
+
+func TestParseSelectorInvalid(t *testing.T) {
+	for _, testcase := range []struct {
+		input string
+	}{
+		{
+			input: "", // empty
+		},
+		{
+			input: "/linux/arm", // leading slash
+		},
+		{
+			input: "linux/arm/", // trailing slash
+		},
+		{
+			input: "linux /arm", // spaces
+		},
+		{
+			input: "linux/&arm", // invalid character
+		},
+		{
+			input: "linux/arm/foo/bar", // invalid character
```

Comment:
minor nit: comment says `invalid character` for this last test when I think this is actually valid parts, but `len(parts) > 3`, so no match on the switch.